spring-expression/src/main/java/org/springframework/expression/spel/support/ReflectionHelper.java

```diff
@@ -457,14 +459,18 @@ private static boolean isFirstEntryInArray(Object value, @Nullable Object possib
 	 * @return a repackaged array of arguments where any varargs setup has been done
 	 */
 	public static Object[] setupArgumentsForVarargsInvocation(Class<?>[] requiredParameterTypes, Object... args) {
-		// Check if array already built for final argument
 		int parameterCount = requiredParameterTypes.length;
+		Assert.notEmpty(requiredParameterTypes, "Required parameter types must not be empty");
+
+		Class<?> lastRequiredParameterType = requiredParameterTypes[parameterCount - 1];
+		Assert.isTrue(lastRequiredParameterType.isArray(), "Method must be varargs");
+
 		int argumentCount = args.length;
+		Object lastArgument = argumentCount > 0 ? args[argumentCount - 1] : null;
 
 		// Check if repackaging is needed...
 		if (parameterCount != args.length ||
-				requiredParameterTypes[parameterCount - 1] !=
-						(args[argumentCount - 1] != null ? args[argumentCount - 1].getClass() : null)) {
+			(!isArray(lastArgument) && differentTypes(lastRequiredParameterType, lastArgument))) {
```

Comment:
This can be simplified by using `java.lang.Class.isInstance(Object)`.

However, there is no need to update this PR: I'll do that when merging.

Reply:
Thank you, I didn't think of that!